### PR TITLE
shevkunov: hw1 - transformers

### PR DESCRIPTION
**Pull request by szewk**
Есть проблемы с Линформером, мне кажется я очень криво реализовал

### CODE REVIEW

File: hw-transformers-shevk.ipynb [hunk 1]
```diff
@@ -0,0 +1 @@
...w5LIGn/yeF6PBAjuY6+d2/NvLZmowW62TSXunds0hzOMOedjUGJ0XxwOS+mEywfOMR3vpyX7MV7txuL397Zxs7DhYQaLPwm9sG0i0upM52Vw1LBGD9zuO+ceabyqY9ueQXVxJi92No72gArBYzlw6KAWBFdUGvKS2v7tnXr2s4cVEOAC4dFIMJ2J6WT3ZeWZOfU0REpL3xuKvmtDP0tkNEREROItfX005z2omItBU9EkL5+88vY3BSFC6Pl3mf7+Uvb24hv7iitZt23tK7Z5FmYKqe0+5Mi3YAF/bvyF0TegPw6bcZfPjVwSZoWW1er8Gcj3eydX8e/n5mfnXrQHokhNa7baeYIPp2CcNrGCxZf7hJ27FsQ9XxRg6OxWr94bJU0/Ptu3255BU13T8Ct8frKwSOqe7ZBxAdFkj/buEAfPVd8/a2c2Xto/Lw7mY9h4iISGurmdPO0PCYIiIichLqaSci0jaFOGz8+taBTKue1mnr/jx+8891bNilKYaag4p2Is2hpmhnGGc0RGaNy5LjuO2qngC89980Pvs2oyla5/P2kn2s3paNxWzi5zdeQO/OYY1uf1VqVW+75RuPUOk684LkibLzyti6Pw+AUUPia62Li3LQu1MohtG0RbQNu3IoKKkkLMjGkN5RtdZdllw1l97K77Karat3xYHNFLz3fxS+P4PyPWub5RwiIiJtgbdmeEyzinYiIiLSsEqXh8KSSkBFOxGRtshkMnHFsAT+7+6hdI4Jotjp4vn523j54504y92t3bzziop2Is3BfMKflvfMi3YA1wxPZMrlXQGY98Ve39xvZ2vxt+l8Wl0EvPsnvRnQI/Kk+yQnRRIVGkBJmYvV27JP63ze8hJc2WmU71mHc/0iSlbNx338kO/5XNA9go4RgXX2qymirdh8pMmKaDU9+y5LjsNqqX0ZHNIrCkeAlbyiCrYdyGuS853IdewgRZ//E6qHOy1e8i9cR/Y2+XlERETaAq9Hc9qJiIjIyeUUVk1R4e9nJijQ2sqtERGRhiR0cPD7n6Uw7qJOAKzYnMWvXlzDxt05rdyy84f+C4o0A9MJQ0BVHtyCUeHEW1qI11n9KC3AZAvAPnQCfrE9T3q8SZd2oazCw6er0/n3J7vYnpZPSq8oBvSIxB5w+n/Gq7dl88YX+wC4aUx3LhkYe0r7WSxmrhgWz9tL9vPF2kxGDo
```
Comment (Pretzandalf) [inline hw-transformers-shevk.ipynb:1]: <p>Забыл докинуть дропаут, все же это сильно поможет бороться с переобучением.</p><p>Возможно тот резкий провал точности валидации на графике связан именно с отсутствием дропаута.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/szewk/summer-ml-2025/pull/1/discussion/#comment-2189708825'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

Comment (Pretzandalf) [inline hw-transformers-shevk.ipynb:1]: <p>Все в этом задании сделано верно, но просто добавлю для уточнения.</p><p>В настоящих моделях клс токен добавляют как параметр модели через</p>

<pre class="ql-syntax" spellcheck="false">self.cls_token&nbsp;= nn.Parameter(torch.randn(embed_dim))
</pre><p>И второй момент.</p>

<pre class="ql-syntax" spellcheck="false">&nbsp;&nbsp;&nbsp;&nbsp;x = self.positional_encoding(self.input_embedding(x))
&nbsp;&nbsp;&nbsp;&nbsp;x = torch.cat((x, self.cls_token.repeat(x.shape[0], 1, 1).cuda()), dim=1)
</pre><p>Можно клс токен добавлять как до pe так и после (и чаще всего после). Например так делают в BERT, но здесь это не важно.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/szewk/summer-ml-2025/pull/1/discussion/#comment-2189708836'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='51'/>

Comment (Pretzandalf) [inline hw-transformers-shevk.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #25.            self.e_proj = nn.Parameter(torch.empty((20, 257), dtype=torch.float32))
</pre><p>Давай без хардкодинга) Вот так нужно было nn.Parameter(torch.empty(proj_dim, sequence_length))</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/szewk/summer-ml-2025/pull/1/discussion/#comment-2189708844'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='79'/>

Comment (Pretzandalf) [inline hw-transformers-shevk.ipynb:1]: <p>Вообще все отлично сделано, не знаю чем тебе не понравилось твоя реализация)</p><p> </p><p>Если тебе не нравится разница графиков с эталонным решением, то тут разница только в том что у тебя эпох в 10 раз больше.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/szewk/summer-ml-2025/pull/1/discussion/#comment-2189708858'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='91'/>

Comment (Pretzandalf) [inline hw-transformers-shevk.ipynb:1]: <p>В итоге:</p><p>Все хорошо сделано. Я добавил только пару уточений, чтобы в будующем не было недопониманий</p><p>Молодец!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/szewk/summer-ml-2025/pull/1/discussion/#comment-2189708870'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='93'/>
